crates/sui-graphql-rpc/src/types/move_package.rs

```diff
@@ -575,6 +629,75 @@ impl Loader<PackageVersionKey> for Db {
     }
 }
 
+#[async_trait::async_trait]
+impl Loader<LatestKey> for Db {
+    type Value = SuiAddress;
+    type Error = Error;
+
+    async fn load(&self, keys: &[LatestKey]) -> Result<HashMap<LatestKey, SuiAddress>, Error> {
+        use packages::dsl;
+        let other = diesel::alias!(packages as other);
+
+        let mut ids_by_cursor: BTreeMap<_, BTreeSet<_>> = BTreeMap::new();
+        for key in keys {
+            ids_by_cursor
+                .entry(key.checkpoint_viewed_at)
+                .or_default()
+                .insert(key.address.into_vec());
+        }
+
+        // Issue concurrent reads for each group of IDs
+        let futures = ids_by_cursor
+            .into_iter()
+            .map(|(checkpoint_viewed_at, ids)| {
+                self.execute(move |conn| {
+                    let results: Vec<(Vec<u8>, Vec<u8>)> = conn.results(|| {
+                        let o_original_id = other.field(dsl::original_id);
+                        let o_package_id = other.field(dsl::package_id);
+                        let o_cp_seq_num = other.field(dsl::checkpoint_sequence_number);
+                        let o_version = other.field(dsl::package_version);
+
+                        let query = dsl::packages
+                            .inner_join(other.on(dsl::original_id.eq(o_original_id)))
+                            .select((dsl::package_id, o_package_id))
+                            .filter(dsl::package_id.eq_any(ids.iter().cloned()))
+                            .filter(o_cp_seq_num.le(checkpoint_viewed_at as i64))
+                            .order_by((dsl::package_id, dsl::original_id, o_version.desc()))
```

Comment:
i assume this is pretty performant already, but what if we order all 3 in descending order? would that further leverage the index?

Reply:
I can give it a try!

Reply:
This doesn't make a big difference, but this comment did lead me to change the primary key of this table to `(package_id, original_id, package_version)` and add a separate uniqueness constraint on `package_id`, and that does reduce the cost slightly.